Realizado por Katalina Morales Dueñas

Codigo: 2262485

31 de Octubre del 2022


# Árbol de decisión

Según el siguiente árbol de decisión, analice y responda:

![Obama-Clinton decision tree](https://github.com/justmarkham/DAT8/raw/226791169b1cc6df8e8845c12e34e748d5ffaa85/notebooks/images/obama_clinton_tree.jpg)

## 1.1 ¿Qué son las observaciones? ¿Cuántas observaciones hay?


Las observaciones corresponden a los condados de Estados Unidos que participaron en una encuesta para elegir a su futuro presidente. No se incluyeron los condados de Florida y Michigan. Hay 2241 observaciones.

## 1.2 ¿Cuál es la variable de respuesta?


La variable respuesta es determinar cuál participante ganará  en el concurso de nominaciones, si Hillary o Barack.
Ej: 1 Hillary 0 Obama 
(De acuerdo a las votaciones por condados)

## 1.3 ¿Cuáles son las características?¿Cuál es la característica más predictiva?


Caracteristicas:

1. Raza: El porcentaje de la población que pertenece a la raza negra que hay en cada condado. (Esta es la característica más predictiva porque es la primera ramificación del arbol, la más relevante)
2. Educación: Tasa de graduación de la escuela secundaria por condado (En porcentaje). Esta caracteristica se evalúa con un 78% y luego es más estricto con un 87%. 
3. Ubicación: Ubicación del condado. (Norte, sur, este, oeste)
4. Pobreza: Promedio del salario por condado. (El 47% de la población por condado ganó menos de 30.000 USD y el 53% ganó más de 30.000 USD).
5. Densidad de la población: Cantidad de personas por milla cuadrada por condado.
6. De acuerdo a elecciones presidenciales pasadas. Condados donde Bush venció Kerry en más de 16,5 puntos porcentuales.

## 1.4 ¿Cuál es la predicción de clase para el siguiente condado con los siguientes datos: 15% de afroamericanos, 90% de tasa de graduación en la escuela secundaria, ubicado en el Sur, alta pobreza, alta densidad de población?


Obama

## 1.5 ¿Cuál es el chance predicha para ese mismo condado anteriormente descrito?

185 a 36

## SVM


Para esta tarea consideramos un conjunto de observaciones sobre una serie de variedades de vino tinto y blanco que implican sus propiedades químicas y su clasificación por parte de los catadores. La industria del vino ha experimentado un gran crecimiento en los últimos tiempos debido al aumento del consumo social. El precio del vino depende de un concepto bastante abstracto de apreciación del vino por parte de los catadores, cuya opinión puede tener un alto grado de variabilidad. El precio del vino depende en cierta medida de este factor tan volátil. Otro factor clave en la certificación y evaluación de la calidad del vino son las pruebas fisicoquímicas, que se realizan en laboratorio y tienen en cuenta factores como la acidez, el nivel de pH, la presencia de azúcar y otras propiedades químicas. Para el mercado del vino, sería interesante que la calidad humana de la cata pudiera relacionarse con las propiedades químicas del vino para que el proceso de certificación y evaluación de la calidad estuviera más controlado.

Se dispone de dos conjuntos de datos, uno de los cuales se refiere al vino tinto y cuenta con 1.599 variedades diferentes, y el otro, al vino blanco, con 4.898 variedades. Todos los vinos se producen en una zona concreta de Portugal. Se recogen datos sobre 12 propiedades diferentes de los vinos, una de las cuales es la calidad, basada en datos sensoriales, y el resto son propiedades químicas de los vinos, como la densidad, la acidez, el contenido de alcohol, etc. Todas las propiedades químicas de los vinos son variables continuas. La calidad es una variable ordinal con una clasificación posible de 1 (peor) a 10 (mejor). Cada variedad de vino es catada por tres catadores independientes y la clasificación final asignada es la mediana de la clasificación dada por los catadores.

Se espera que un modelo predictivo desarrollado a partir de estos datos sirva de orientación a los viñedos en cuanto a la calidad y el precio que se espera obtener de sus productos sin depender en gran medida de la volatilidad de los catadores.


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data_r = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_red.csv')
data_w = pd.read_csv('https://raw.githubusercontent.com/prof-tatiana-gutierrez/DatosMD/master/Wine_data_white.csv')

In [ ]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36            20.7      0.045   
1               6.3             0.300         0.34             1.6      0.049   
2               8.1             0.280         0.40             6.9      0.050   
3               7.2             0.230         0.32             8.5      0.058   
4               7.2             0.230         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
6492            6.2             0.600         0.08             2.0      0.090   
6493            5.9             0.550         0.10             2.2      0.062   
6494            6.3             0.510         0.13             2.3      0.076   
6495            5.9             0.645         0.12             2.0      0.075   
6496            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
6492                 32.0                  44.0  0.99490  3.45       0.58   
6493                 39.0                  51.0  0.99512  3.52       0.76   
6494                 29.0                  40.0  0.99574  3.42       0.75   
6495                 32.0                  44.0  0.99547  3.57       0.71   
6496                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality   type  
0         8.8        6  white  
1         9.5        6  white  
2        10.1        6  white  
3         9.9        6  white  
4         9.9        6  white  
...       ...      ...    ...  
6492     10.5        5    red  
6493     11.2        6    red  
6494     11.0        6    red  
6495     10.2        5    red  
6496     11.0        6    red  

[6497 rows x 13 columns]

In [ ]:
data.quality.value_counts()

6    2836
5    2138
7    1079
4     216
8     193
3      30
9       5
Name: quality, dtype: int64

In [ ]:
data.type.value_counts()

white    4898
red      1599
Name: type, dtype: int64

# Ejercicio 2.1

Mostrar la tabla de frecuencias de la calidad por tipo de vino.

Vino rojo

In [ ]:
(data_r 
  .groupby("quality")
  .agg(frequency=("quality", "count")))

frequency
quality           
3               10
4               53
5              681
6              638
7              199
8               18

Vino Blanco

In [ ]:
(data_w 
  .groupby("quality")
  .agg(frequency=("quality", "count")))

frequency
quality           
3               20
4              163
5             1457
6             2198
7              880
8              175
9                5

# Ejercicio 2.2

Teniendo en cuenta todas las variables:

* Crear un objetivo binario para cada tipo de vino, haciendo la transformación necesaria para ingresar la información al modelo.
* Crear una SVM que prediga el tipo de vino
* Evalúe con F1, AUC-ROC y log-loss


In [ ]:
from sklearn.preprocessing import label_binarize
#Crear un objetivo binario para cada tipo de vino, haciendo la transformación necesaria para ingresar la información al modelo.
X = np.array(data.drop(['type'], 1))
y = np.array(data['type'])
y = label_binarize(y, classes=["red","white"])

In [ ]:
from sklearn import preprocessing, metrics, model_selection
import pandas as pd
from sklearn.metrics import roc_auc_score, accuracy_score, make_scorer, roc_curve, auc
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

In [ ]:
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.3)

In [ ]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
#Crear una SVM que prediga el tipo de vino
from sklearn.svm import SVC
modSVC = SVC(kernel='linear', probability = True)
modSVC.fit(X_train, y_train)

SVC(kernel='linear', probability=True)

In [ ]:
y_pred = modSVC.predict(X_test);y_pred 

array([1, 1, 1, ..., 1, 1, 0])

In [ ]:
print("Cantidad de vinos reales que son blancos:", y_test.sum())
print("Cantidad de vinos estimados que son blancos:", y_pred.sum())

Cantidad de vinos reales que son blancos: 1461
Cantidad de vinos estimados que son blancos: 1462


In [ ]:
print("Cantidad de vinos reales que son rojos:", len(y_test)-y_test.sum())
print("Cantidad de vinos estimados que son rojos:",len(y_pred)- y_pred.sum())

Cantidad de vinos reales que son rojos: 489
Cantidad de vinos estimados que son rojos: 488


In [ ]:
y_pred_prob = modSVC.predict_proba(X_test); y_pred_prob

array([[2.07531135e-09, 9.99999998e-01],
       [2.75705357e-09, 9.99999997e-01],
       [3.26366882e-01, 6.73633118e-01],
       ...,
       [1.86401752e-05, 9.99981360e-01],
       [9.94987502e-01, 5.01249757e-03],
       [6.73432113e-03, 9.93265679e-01]])

In [ ]:
# calculate log loss
from sklearn import metrics
print(metrics.log_loss(y_test, y_pred_prob))

0.056433488017590676


In [ ]:
#Calcular F1 
print(metrics.f1_score(y_test, y_pred))

0.991216216216216


In [ ]:
#AUC-ROC 
print(metrics.roc_auc_score(y_test, y_pred))

0.9803889245659503


# Ejercicio 2.3

* Probar la SVM utilizando los diferentes kernels ('poly', 'rbf', 'sigmoid')
* Evalúe con F1, AUC-ROC y log-loss


POLY

In [ ]:
from sklearn.svm import SVC
modSVC = SVC(kernel='poly', probability = True)
modSVC.fit(X_train, y_train)

SVC(kernel='poly', probability=True)

In [ ]:
#Evalúe con F1, AUC-ROC y log-loss

In [ ]:
y_pred_prob = modSVC.predict_proba(X_test)
y_pred = modSVC.predict(X_test)

In [ ]:
print("Cantidad de vinos reales que son blancos:", y_test.sum())
print("Cantidad de vinos estimados que son blancos:", y_pred.sum())

Cantidad de vinos reales que son blancos: 1461
Cantidad de vinos estimados que son blancos: 1447


In [ ]:
print("Cantidad de vinos reales que son rojos:", len(y_test)-y_test.sum())
print("Cantidad de vinos estimados que son rojos:",len(y_pred)- y_pred.sum())

Cantidad de vinos reales que son rojos: 489
Cantidad de vinos estimados que son rojos: 503


In [ ]:
print("Log loss: ", metrics.log_loss(y_test, y_pred_prob))
print("F1: ", metrics.f1_score(y_test, y_pred))
print("AUC-ROC: ", metrics.roc_auc_score(y_test, y_pred))

Log loss:  0.255097450380746
F1:  0.9461199593358185
AUC-ROC:  0.8906920139446343


RBF

In [ ]:
from sklearn.svm import SVC
modSVC_2 = SVC(kernel='rbf', probability = True)
modSVC_2.fit(X_train, y_train)

SVC(probability=True)

In [ ]:
y_pred_prob_2 = modSVC_2.predict_proba(X_test)
y_pred_2 = modSVC_2.predict(X_test)

In [ ]:
print("Cantidad de vinos reales que son blancos:", y_test.sum())
print("Cantidad de vinos estimados que son blancos:", y_pred_2.sum())

Cantidad de vinos reales que son blancos: 1461
Cantidad de vinos estimados que son blancos: 1524


In [ ]:
print("Cantidad de vinos reales que son rojos:", len(y_test)-y_test.sum())
print("Cantidad de vinos estimados que son rojos:",len(y_pred_2)- y_pred_2.sum())

Cantidad de vinos reales que son rojos: 489
Cantidad de vinos estimados que son rojos: 426


In [ ]:
#Evalúe con F1, AUC-ROC y log-loss

In [ ]:
print("Log loss: ", metrics.log_loss(y_test, y_pred_prob_2))
print("F1: ", metrics.f1_score(y_test, y_pred_2))
print("AUC-ROC: ", metrics.roc_auc_score(y_test, y_pred_2))

Log loss:  0.21850691786111306
F1:  0.9567839195979899
AUC-ROC:  0.8905468563006262


SIGMOID

In [ ]:
from sklearn.svm import SVC
modSVC_3 = SVC(kernel='sigmoid', probability = True)
modSVC_3.fit(X_train, y_train)

SVC(kernel='sigmoid', probability=True)

In [ ]:
y_pred_prob_3 = modSVC_3.predict_proba(X_test)
y_pred_3 = modSVC_3.predict(X_test)

In [ ]:
print("Cantidad de vinos reales que son blancos:", y_test.sum())
print("Cantidad de vinos estimados que son blancos:", y_pred_3.sum())

Cantidad de vinos reales que son blancos: 1461
Cantidad de vinos estimados que son blancos: 1464


In [ ]:
print("Cantidad de vinos reales que son rojos:", len(y_test)-y_test.sum())
print("Cantidad de vinos estimados que son rojos:",len(y_pred_3)- y_pred_3.sum())

Cantidad de vinos reales que son rojos: 489
Cantidad de vinos estimados que son rojos: 486


In [ ]:
#Evalúe con F1, AUC-ROC y log-loss

In [ ]:
print("Log loss: ", metrics.log_loss(y_test, y_pred_prob_3))
print("F1: ", metrics.f1_score(y_test, y_pred_3))
print("AUC-ROC: ", metrics.roc_auc_score(y_test, y_pred_3))

Log loss:  0.48354128067064334
F1:  0.8403418803418803
AUC-ROC:  0.680316028604662


¡¡Tenemos un ganador!!

Kernel = RBF

Métricas:
- Log loss:  0.21850691786111306
- F1:  0.9567839195979899
- AUC-ROC:  0.8905468563006262

Es curioso porque si verificamos con el grupo de testeo la cantidad de vinos que están bien clasificados con respecto al grupo real son mucho menos a comparación del kernel SIGMOUND (El cual lo clasifica mejor pero tiene las peores métricas.) Además de ello con un Kernel lineal las estimaciones dan muy aproximidas a las reales, sin embargo, puede ocasionar overfitting.

# Ejercicio 2.4
* Utilizando el mejor SVM del punto anterior encontrar los parámetros que da el mejor rendimiento teniendo en cuenta los siguientes hiperparámetros
'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]
* Evalúe cada modelo con F1, AUC-ROC y log-loss

Nota: Para los valores de C y gamma, deben realizar todas las posibles combinaciones entre estas dos variables.

In [ ]:
c_range = [0.1, 1, 10, 100, 1000]
gamma_range = [0.01, 0.001, 0.0001]
import itertools
for combinacion in itertools.product(c_range,gamma_range):
    print(combinacion)

(0.1, 0.01)
(0.1, 0.001)
(0.1, 0.0001)
(1, 0.01)
(1, 0.001)
(1, 0.0001)
(10, 0.01)
(10, 0.001)
(10, 0.0001)
(100, 0.01)
(100, 0.001)
(100, 0.0001)
(1000, 0.01)
(1000, 0.001)
(1000, 0.0001)


Log loss

In [ ]:
log_loss_1 = []
c_range = [0.1,0.1,0.1,1,1,1,10,10,10,100,100,100,1000,1000,1000]
gamma_range = [0.01, 0.001, 0.0001, 0.01, 0.001, 0.0001, 0.01, 0.001, 0.0001, 0.01, 0.001, 0.0001, 0.01, 0.001, 0.0001]
j=0
for i in c_range: 
  modSVC_prueba = SVC(kernel='rbf', C = i, gamma = gamma_range[j], probability = True)
  modSVC_prueba.fit(X_train, y_train)
  y_pred_prob = modSVC_prueba.predict_proba(X_test)
  log_loss_1.append(metrics.log_loss(y_test, y_pred_prob))
  j=j+1
log_loss_1 

[0.21621739689550856,
 0.22308916911502988,
 0.229490152029954,
 0.16913547879247223,
 0.1957960329834863,
 0.20829134960043452,
 0.08971921735456767,
 0.14054064635981567,
 0.16490928444752645,
 0.06635005743489227,
 0.05420879297820602,
 0.10436543097545009,
 0.06840131904166924,
 0.037812464078783865,
 0.051596536108423345]

F1

In [ ]:
F_1 = []
j=0
for i in c_range: 
  modSVC_prueba = SVC(kernel='rbf', C = i, gamma = gamma_range[j], probability = True)
  modSVC_prueba.fit(X_train, y_train)
  y_pred = modSVC_prueba.predict(X_test)
  F_1.append(metrics.f1_score(y_test, y_pred))
  j=j+1
F_1  

[0.9589675711449371,
 0.9575033200531209,
 0.9538971807628523,
 0.9664340312396145,
 0.9615384615384615,
 0.9592039800995025,
 0.9828224991579656,
 0.9716193656093488,
 0.9651510122801196,
 0.9868376645291934,
 0.9898717083051991,
 0.9739478957915831,
 0.9877966101694916,
 0.99322951929587,
 0.9901927629354076]

AUC - ROC

In [ ]:
AUC = []
j=0
for i in c_range: 
  modSVC_prueba = SVC(kernel='rbf', C = i, gamma = gamma_range[j], probability = True)
  modSVC_prueba.fit(X_train, y_train)
  y_pred = modSVC_prueba.predict(X_test)
  AUC.append(metrics.roc_auc_score(y_test, y_pred))
  j=j+1
AUC  

[0.8895538519758721,
 0.8903637531249283,
 0.8815953189147037,
 0.9103131235522122,
 0.8973074012063944,
 0.892776255131763,
 0.9596740900438063,
 0.9265412490539208,
 0.9060758354166188,
 0.9702228733285935,
 0.9768726634710213,
 0.9317934221692162,
 0.9770059746336094,
 0.9867448567693402,
 0.979374039586248]

# Ejercicio 2.5

Compare los resultados con todos los anteriores modelos y eliga ¿Cuál algoritmo con cuáles hiperparámetros es el que tiene mejor rendimiento según la puntuación F1, AUC-ROC y log-loss?

Tabla con las métricas

In [ ]:
tabla=np.array([log_loss_1,
               F_1,
               AUC, gamma_range])
frame=pd.DataFrame(tabla.T, index=c_range,columns=['Log loss','F_1 Score','AUC - ROC', "Gamma"])
frame

Log loss  F_1 Score  AUC - ROC   Gamma
0.1     0.216217   0.958968   0.889554  0.0100
0.1     0.223089   0.957503   0.890364  0.0010
0.1     0.229490   0.953897   0.881595  0.0001
1.0     0.169135   0.966434   0.910313  0.0100
1.0     0.195796   0.961538   0.897307  0.0010
1.0     0.208291   0.959204   0.892776  0.0001
10.0    0.089719   0.982822   0.959674  0.0100
10.0    0.140541   0.971619   0.926541  0.0010
10.0    0.164909   0.965151   0.906076  0.0001
100.0   0.066350   0.986838   0.970223  0.0100
100.0   0.054209   0.989872   0.976873  0.0010
100.0   0.104365   0.973948   0.931793  0.0001
1000.0  0.068401   0.987797   0.977006  0.0100
1000.0  0.037812   0.993230   0.986745  0.0010
1000.0  0.051597   0.990193   0.979374  0.0001

In [ ]:
frame["Log loss"].min() #Buscamos la pérdida logarítmica más pequeña

0.037812464078783865

In [ ]:
frame["F_1 Score"].max() #Buscamos el F1 Score más grande

0.99322951929587

In [ ]:
frame["AUC - ROC"].max() #Buscamos el AUC - ROC más grande

0.9867448567693402

De acuerdo a las tres métricas utilizadas (Log Loss, F1 Score y AUC-ROC) el modelo kernel = "RBF" que tiene el mejor rendimiento es utilizando los siguientes valores en los hiperparámetros:
- C = 1000
- Gamma = 0.001

In [ ]:
modSVC_prueba = SVC(kernel='rbf', C = 1000, gamma = 0.001, probability = True)
modSVC_prueba.fit(X_train, y_train)
y_pred_4 = modSVC_prueba.predict(X_test)

In [ ]:
print("Cantidad de vinos reales que son blancos:", y_test.sum())
print("Cantidad de vinos estimados que son blancos:", y_pred_4.sum())

Cantidad de vinos reales que son blancos: 1461
Cantidad de vinos estimados que son blancos: 1454


In [ ]:
print("Cantidad de vinos reales que son rojos:", len(y_test)-y_test.sum())
print("Cantidad de vinos estimados que son rojos:",len(y_pred_4)- y_pred_4.sum())

Cantidad de vinos reales que son rojos: 489
Cantidad de vinos estimados que son rojos: 496
